src/Microsoft.DotNet.XHarness.Android/InstrumentationRunner.cs

```diff
@@ -0,0 +1,210 @@
+        }
+
+        if (resultValues.TryGetValue(ShortMessageVariableName, out string? shortMessage))
+        {
+            _logger.LogInformation($"Short message:{Environment.NewLine}{shortMessage}");
+            processCrashed = shortMessage.Contains(ProcessCrashedShortMessage);
+        }
+
+        // Due to the particulars of how instrumentations work, ADB will report a 0 exit code for crashed instrumentations
+        // We'll change that to a specific value and print a message explaining why.
+        int? instrumentationExitCode = null;
+        if (resultValues.TryGetValue(ReturnCodeVariableName, out string? returnCode))
+        {
+            if (int.TryParse(returnCode, out int parsedExitCode))
+            {
+                _logger.LogInformation($"Instrumentation finished normally with exit code {parsedExitCode}");
+                instrumentationExitCode = parsedExitCode;
+            }
+            else
+            {
+                _logger.LogError($"Un-parse-able value for '{ReturnCodeVariableName}' : '{returnCode}'");
+            }
+        }
+        else
+        {
+            _logger.LogError($"No value for '{ReturnCodeVariableName}' provided in instrumentation result. This may indicate a crashed test (see log)");
+        }
+
+        return (instrumentationExitCode, processCrashed, failurePullingFiles);
+    }
+
+    private bool PullResultXMLs(string apkPackageName, string outputDirectory, Dictionary<string, string> resultValues)
+    {
+        bool success = false;
+
+        foreach (string possibleResultKey in s_xmlOutputVariableNames)
+        {
+            if (!resultValues.TryGetValue(possibleResultKey, out string? resultFile))
+            {
+                continue;
+            }
+
+            _logger.LogInformation($"Found XML result file: '{resultFile}'(key: {possibleResultKey})");
+
+            try
+            {
+                _runner.PullFiles(apkPackageName, resultFile, outputDirectory);
+            }
+            catch (Exception toLog)
+            {
+                _logger.LogError(toLog, "Hit error (typically permissions) trying to pull {filePathOnDevice}", resultFile);
+                success = true;
+            }
+        }
+
+        return success;
+    }
+
+    private Dictionary<string, string> ParseInstrumentationOutputs(string stdout)
+    {
+        var outputs = new Dictionary<string, string>();
+        string[] lines = stdout.Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries);
+
+        foreach (string line in lines.Where(line => line.StartsWith(InstrumentationResultPrefix)))
+        {
+            var subString = line.Substring(InstrumentationResultPrefix.Length);
+            string[] results = subString.Trim().Split('=');
+            if (results.Length == 2)
+            {
+                if (outputs.ContainsKey(results[0]))
```

Comment:
Probably that's a common knowledge but for someone not familiar with the format of the data it's not clear what is `result[0]` and `result[1]` are. You could assign them to variables with some descriptive names or at least add a comment describing the line format.